lib/routes/account.js

```diff
@@ -660,34 +534,48 @@ module.exports = (log, db, mailer, Password, config, customs, checkPassword, pus
           // login session before they can actually use it. Otherwise, they don't *have* to verify
           // their session. All sessions are created unverified because it prevents them from being
           // used for sync.
-          mustVerifySession = needsVerificationId && (requestHelper.wantsKeys(request) || verificationMethod)
-
-          // If the email itself is unverified, we'll re-send the "verify your account email" and
-          // that will suffice to confirm the sign-in.  No need for a separate confirmation email.
-          doSigninConfirmation = mustVerifySession && accountRecord.primaryEmail.isVerified
-
-          let flowCompleteSignal
-          if (service === 'sync') {
-            flowCompleteSignal = 'account.signed'
-          } else if (doSigninConfirmation) {
-            flowCompleteSignal = 'account.confirmed'
-          } else {
-            flowCompleteSignal = 'account.login'
-          }
-          request.setMetricsFlowCompleteSignal(flowCompleteSignal, 'login')
+          const mustVerifySession = needsVerificationId && (requestHelper.wantsKeys(request) || verificationMethod)
 
-          return checkPassword(accountRecord, authPW, request.app.clientAddress)
-            .then(
-              function (match) {
-                if (! match) {
-                  throw error.incorrectPassword(accountRecord.email, email)
-                }
+          return P.resolve()
+            .then(() => {
+              if (! needsVerificationId) {
+                return []
+              }
+              return [random.hex(16), TokenCode()]
+            })
+            .spread((tokenVerificationId, tokenVerificationCode) => {
```

Comment:
TIL that you don't need to use `P.all` in order to spread promises, if you're already inside a promise chain. 😄